# **Functions Assingment **

**Theory Questions:**


# **Q1. What is the difference between a function and a method in Python?**

Difference between Function and Method in Python:

**Function**

A function is a block of reusable code that is defined using the def keyword.

It can exist independently (outside of a class).

It is called directly using its name.

**Method**

A method is also a function, but it is defined inside a class.

It is associated with an object (instance of a class).

It is called using the dot (.) operator on an object.


**Function: **Defined outside a class, called directly.

Method: Defined inside a class, called using an object.

 **Example:**

In [ ]:
# Function (independent)
def greet(name):
    return f"Hello, {name}!"

print(greet("Neha"))   # Calling a function


# Method (inside a class)

class Person:
    def __init__(self, name):
        self.name = name

    # This is a method
    def greet(self):
        return f"Hello, {self.name}!"


p = Person("Neha")
print(p.greet())   # Calling a method using object

Hello, Neha!
Hello, Neha!


# **Q2. Explain the concept of function arguments and parameters in Python.**

In Python, parameters and arguments are related but slightly different concepts:

Parameters are the variables defined inside the function header. They act as placeholders for the values the function will receive.

Arguments are the actual values or data you pass to a function when calling it.


**Parameters** = variables inside the function definition

**Arguments **= values passed during function call

**Example1:**

In [ ]:
# Function definition with parameters
def greet(name, age):
    print(f"Hello {name}, you are {age} years old.")

# Function call with arguments
greet("Neha", 21)

Hello Neha, you are 21 years old.


Here, name and age are parameters (defined in the function).

"Neha" and 21 are arguments (passed during the function call).

# **Q3. What are the different ways to define and call a function in Python?**

In Python, parameters and arguments are related but slightly different concepts:

Parameters → Variables that are defined inside the function definition. They act as placeholders for the values the function will receive.

**Arguments** → Actual values or data you pass to the function when calling it.

**Example 1:**

In [ ]:
# Function definition with parameters

def greet(name, age):
    print(f"Hello {name}, you are {age} years old.")

# Function call with arguments
greet("Neha", 21)

Hello Neha, you are 21 years old.


name and age → parameters (defined in the function).

"Neha" and 21 → arguments (values passed when calling the function).

So, parameters are like empty boxes in the function, and arguments are the actual items you put into those boxes when using the function.

# **Q4. What is the purpose of the `return` statement in a Python function?**

The return statement in a Python function is used to send a value (or multiple values) back to the caller.
It ends the functions execution and provides the result that the function computes.

If you dont use return, the function will return None by default.


Example1:

In [ ]:
def add_numbers(a, b):

    result = a + b
    return result  # sends the result back to the caller

# Calling the function
sum_value = add_numbers(5, 7)
print("The sum is:", sum_value)
sum_value = add_numbers(5, 7)
print("The sum is:", sum_value)

The sum is: 12
The sum is: 12


Here, the return statement sends the value 12 back, which gets stored in sum_value.

# **Q5 What are iterators in Python and how do they differ from iterables?**

Iterable

An iterable is any Python object capable of returning its members one at a time.

Examples: list, tuple, string, set, dictionary.

It implements the __iter__() method, which returns an iterator.

Iterator

An iterator is an object that actually goes through (iterates) the elements of an iterable one at a time.

It implements both:

__iter__() (returns itself)

__next__() (returns the next item, or raises StopIteration when done).


Iterable = a collection you can loop over.

Iterator = the mechanism that fetches elements from the iterable one by one.

In [ ]:
# 1 Example


# An iterable
my_list = [10, 20, 30]

# Converting iterable into an iterator
my_iterator = iter(my_list)

# Accessing elements using iterator
print(next(my_iterator))  # 10
print(next(my_iterator))  # 20
print(next(my_iterator))  # 30

# If we call again, it will raise StopIteration
# print(next(my_iterator))

10
20
30


my_list is iterable.

my_iterator is an iterator that goes through elements one by one.

# **Q6. Explain the concept of generators in Python and how they are defined.**

How Generators are Defined

There are two main ways to define a generator:

Using a function with the yield keyword

A normal function that contains yield becomes a generator function.

Instead of returning all results at once, it yields one result at a time.

Each call to next() resumes execution from where it left off.

Using generator expressions (similar to list comprehensions but with () instead of []).

Example 1: Generator Function

In [ ]:
def number_generator(n):
    for i in range(1, n+1):
        yield i   # yields values one by one

gen = number_generator(5)

for num in gen:
    print(num)

1
2
3
4
5


The function doesn’t return a list; instead, it produces numbers one at a time.

This saves memory compared to returning [1, 2, 3, 4, 5].

# **Q7. What are the advantages of using generators over regular functions?**


A lambda function in Python is a small, anonymous (unnamed) function defined using the keyword lambda.
It can have any number of arguments but only one expression, and it automatically returns the result of that expression.

## **Syntax:**

**lambda arguments**: expression

**When to use**:

When you need a short, throwaway function.

Often used with functions like map(), filter(), and sorted().

Useful when defining a function inline instead of writing a full def block.

In [ ]:
# Lambda function to add two numbers
add = lambda x, y: x + y

print(add(5, 3))  # Output: 8

8


Here, lambda x, y: x + y creates a quick function that adds two numbers without using def.

# **Q8. What is a lambda function in Python and when is it typically used?**

A lambda function in Python is a small, anonymous function defined using the keyword lambda. Unlike regular functions created with def, a lambda function is typically used for short, simple operations where defining a full function would be unnecessary.

**Syntax:**

**lambda arguments:** expression


It can take any number of arguments but has only one expression.

The result of the expression is returned automatically.

Typical uses of lambda functions:

As a short helper function inside higher-order functions like map(), filter(), or sorted().

When you need a quick function for a short duration.

In [ ]:
# Example1:

# Using lambda with sorted() to sort by the second element in a tuple
pairs = [(1, 'apple'), (3, 'banana'), (2, 'cherry')]

# Sort based on the number (first element)
sorted_pairs = sorted(pairs, key=lambda x: x[0])

print(sorted_pairs)

[(1, 'apple'), (2, 'cherry'), (3, 'banana')]


Here, lambda x: x[0] is a lambda function that extracts the first element of each tuple for sorting.

# **9. Explain the purpose and usage of the `map()` function in Python**



The map() function in Python is used to apply a function to each item in an iterable (like a list, tuple, or set) and return a map object (an iterator).

Purpose

It helps avoid writing explicit loops.

Makes code more concise and readable.

Useful when the same operation needs to be applied to all elements of a sequence.

Syntax
map(function, iterable)


function: A function to apply to each element.

iterable: The data structure (list, tuple, etc.) whose items will be processed

function: A function to apply to each element.

iterable: The data structure (list, tuple, etc.) whose items will be processed.

In [ ]:
# Example1

# Square each number in a list using map()
numbers = [1, 2, 3, 4, 5]

squared = map(lambda x: x**2, numbers)

print(list(squared))

[1, 4, 9, 16, 25]


Here, lambda x: x**2 is applied to each element of numbers, and map() returns an iterator which is converted into a list.

**Q10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?**

# **Function**

# **Purpose**

**map()** :  Applies a function to each item of an iterable.

**filter()** : Applies a function that returns True/False to each item, keeping only items where the function is True.

**reduce() (from functools)** : Repeatedly applies a function to the elements, reducing them to a single cumulative value.


# **Returns**


**map()** : A new iterable (map object) with transformed elements.

**filter()**: A new iterable (filter object) with filtered elements.

**reduce() (from functools)** :A single value.


# **Typical Use**

**map()  :**Transforming data (e.g., squaring numbers).
**filter():** Filtering data (e.g., keeping only even numbers).

**reduce() (from functools) :** Aggregation (e.g., sum, product, max).



In [ ]:
  # Example with all three
from functools import reduce

numbers = [1, 2, 3, 4, 5]

# 1. map() → square each number
squared = list(map(lambda x: x**2, numbers))
print("map result:", squared)
# [1, 4, 9, 16, 25]

# 2. filter() → keep only even numbers
evens = list(filter(lambda x: x % 2 == 0, numbers))
print("filter result:", evens)
# [2, 4]

# 3. reduce() → calculate product of all numbers
product = reduce(lambda x, y: x * y, numbers)
print("reduce result:", product)
# 120


map result: [1, 4, 9, 16, 25]
filter result: [2, 4]
reduce result: 120


map() → transform each element

filter() → keep some elements based on condition

reduce() → combine all elements into one result

# **11. Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given**
# list:[47,11,42,13]

In [ ]:
from functools import reduce
numbers = [47, 11, 42, 13]
result = reduce(lambda x, y: x + y, numbers)

# Answer attached differently for this question above in a seperate file

## **Answer attached differently for this question above in a seperate file **

# **Practical Questions:**

# **1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in**
the list.

In [ ]:
def sum_even_numbers(numbers):
    total = 0
    for num in numbers:
        if num % 2 == 0:  # Check if the number is even
            total += num
    return total

  # Example usage:
my_list = [1, 2, 3, 4, 5, 6]
result = sum_even_numbers(my_list)
print("Sum of even numbers:", result)

Sum of even numbers: 12


# **2. Create a Python function that accepts a string and returns the reverse of that string.**

In [ ]:
def reverse_string(s):
    return s[::-1]

# Example usage:
input_string = "Hello Earth"
reversed_string = reverse_string(input_string)
print("Reversed string:", reversed_string)

Reversed string: htraE olleH


s[::-1] uses Python slicing to reverse the string.

: → take the whole string

-1 → step backwards, effectively reversing it

# **3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.**

In [ ]:
def square_list(numbers):
    """
    Takes a list of integers and returns a new list
    containing the squares of each number.
    """
    return [num ** 2 for num in numbers]

# Example usage
nums = [1, 2, 3, 4, 5]
print(square_list(nums))

[1, 4, 9, 16, 25]


This uses list comprehension for clean and efficient code.

# **4. Write a Python function that checks if a given number is prime or not from 1 to 200.**

In [ ]:
def is_prime(num):
    """Check if a number is prime."""
    if num < 2:   # 0 and 1 are not prime
        return False
    for i in range(2, int(num**0.5) + 1):  # check divisibility up to sqrt(num)
        if num % i == 0:
            return False
    return True

# Check numbers from 1 to 200
for n in range(1, 201):
    if is_prime(n):
        print(f"{n} is a prime number")
    else:
        print(f"{n} is not a prime number")

1 is not a prime number
2 is a prime number
3 is a prime number
4 is not a prime number
5 is a prime number
6 is not a prime number
7 is a prime number
8 is not a prime number
9 is not a prime number
10 is not a prime number
11 is a prime number
12 is not a prime number
13 is a prime number
14 is not a prime number
15 is not a prime number
16 is not a prime number
17 is a prime number
18 is not a prime number
19 is a prime number
20 is not a prime number
21 is not a prime number
22 is not a prime number
23 is a prime number
24 is not a prime number
25 is not a prime number
26 is not a prime number
27 is not a prime number
28 is not a prime number
29 is a prime number
30 is not a prime number
31 is a prime number
32 is not a prime number
33 is not a prime number
34 is not a prime number
35 is not a prime number
36 is not a prime number
37 is a prime number
38 is not a prime number
39 is not a prime number
40 is not a prime number
41 is a prime number
42 is not a prime number
43 is a pri

This function:

Returns True if a number is prime.

Returns False otherwise.

Then we loop through numbers 1 to 200 and print whether each is prime or not

# **5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.**

In [ ]:
class FibonacciIterator:
    def __init__(self, num_terms):
        self.num_terms = num_terms  # total terms to generate
        self.count = 0              # counter for terms generated
        self.a, self.b = 0, 1       # first two Fibonacci numbers

    def __iter__(self):
        return self

    def __next__(self):
        if self.count >= self.num_terms:
            raise StopIteration  # stop when terms are exhausted
        if self.count == 0:
            self.count += 1
            return self.a
        elif self.count == 1:
            self.count += 1
            return self.b
        else:
            self.a, self.b = self.b, self.a + self.b
            self.count += 1
            return self.b


# Example usage:
fib = FibonacciIterator(10)
for num in fib:
    print(num, end=" ")

0 1 1 2 3 5 8 13 21 34 

# **6. Write a generator function in Python that yields the powers of 2 up to a given**

In [ ]:
# Python generator function that yields the powers of 2 up to a given exponent:

def powers_of_two(n):
    """Generator that yields powers of 2 from 2^0 up to 2^n."""
    for i in range(n + 1):
        yield 2 ** i

# Example usage:
for value in powers_of_two(5):
    print(value)

1
2
4
8
16
32


# **7. Implement a generator function that reads a file line by line and yields each line as a string**

In [ ]:

def read_file_lines(filepath):
    """
    A generator function that reads a file line by line and yields each line as a string.

    Args:
        filepath (str): The path to the file to be read.

    Yields:
        str: Each line from the file, including the newline character if present.
    """
    try:
        with open(filepath, 'r') as file:
            for line in file:
                yield line
    except FileNotFoundError:
        print(f"Error: File not found at {filepath}")
    except Exception as e:
        print(f"An error occurred: {e}")


# 1. Create a dummy file for demonstration
with open("sample.txt", "w") as f:
    f.write("This is line 1.\n")
    f.write("This is line 2.\n")
    f.write("This is line 3.\n")

# 2. Use the generator function to read the file
print("Reading 'sample.txt' using the generator:")
for line in read_file_lines("sample.txt"):
    print(line.strip())  # .strip() removes leading/trailing whitespace, including newline characters

print("\nReading a non-existent file:")
for line in read_file_lines("non_existent_file.txt"):
    print(line.strip())

Reading 'sample.txt' using the generator:
This is line 1.
This is line 2.
This is line 3.

Reading a non-existent file:
Error: File not found at non_existent_file.txt


# **8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.**

In [1]:
# Sample list of tuples
data = [(1, 5), (3, 1), (4, 6), (2, 3), (7, 2)]

# Sort by second element of each tuple
sorted_data = sorted(data, key=lambda x: x[1])

print(sorted_data)

[(3, 1), (7, 2), (2, 3), (1, 5), (4, 6)]


**Explanation:**

lambda x: x [1] → tells Python to use the second element of each tuple as the sorting key.

sorted() → returns a new sorted list.

# **9. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.**

In [ ]:
# Function to convert Celsius to Fahrenheit
def celsius_to_fahrenheit(celsius):
    return (celsius * 9/5) + 32

# List of temperatures in Celsius
celsius_temps = [0, 20, 37, 100]

# Use map() to convert to Fahrenheit
fahrenheit_temps = list(map(celsius_to_fahrenheit, celsius_temps))

# Print results
print("Celsius:    ", celsius_temps)
print("Fahrenheit: ", fahrenheit_temps)

Celsius:     [0, 20, 37, 100]
Fahrenheit:  [32.0, 68.0, 98.6, 212.0]


# **10. Create a Python program that uses `filter()` to remove all the vowels from a given string.**

In [ ]:
def remove_vowels(s):
    vowels = "aeiouAEIOU"
    # filter keeps only characters that are not vowels
    result = ''.join(filter(lambda ch: ch not in vowels, s))
    return result

# Example usage
text = "Hello World"
print("Original String:", text)
print("Without Vowels:", remove_vowels(text))


Original String: Hello World
Without Vowels: Hll Wrld


11) Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:


order no    quantity  price per item   book titel and autho

34587          4         40.95          Learning Python mark Lutz
98762          5         56.80          programming python MARK lutz
77226          3         32.95          Head First python ,Paul Barry
88112          3         24.99          einfuhrung in python3 bernd kein




Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the
product of the price per item and the quantity. The product should be increased by 10,- € if the value of the
order is smaller than 100,00 €.

Write a Python program using lambda and map **bold text**

In [ ]:
# List of orders
orders = [
    [34587, 4, 40.95, "Learning Python mark Lutz"],
    [98762, 5, 56.80, "programming python MARK lutz"],
    [77226, 3, 32.95, "Head First python ,Paul Barry"],
    [88112, 3, 24.99, "einfuhrung in python3 bernd kein"]
]

# Using map and lambda to calculate total with condition
result = list(map(lambda order: (order[0], order[1]*order[2] + 10 if order[1]*order[2] < 100 else order[1]*order[2]), orders))

# Print the result
print(result)

[(34587, 163.8), (98762, 284.0), (77226, 108.85000000000001), (88112, 84.97)]


Notice that the last order gets an extra 10 € because 3 × 24.99 = 74.97 < 100 → 74.97 + 10 = 84.97.